File: src/c04_professional_classes/open_closed/OpenClosedPrinciple.java
package c04_professional_classes.open_closed;

import java.util.List;

/*
In this example:
- AreaCalculator has to know about every shape type, checking the type string to calculate the area.
- This design violates the Open/Closed Principle because adding new shapes (e.g., Triangle) would require modifying calculateArea, leading to less maintainable code.
*/

// Rectangle class representing a rectangle shape
class Rectangle {
    double width;
    double height;

    public Rectangle(double width, double height) {
        this.width = width;
        this.height = height;
    }
}

// Circle class representing a circle shape
class Circle {
    double radius;

    public Circle(double radius) {
        this.radius = radius;
    }
}

// AreaCalculator class responsible for calculating the area of different shapes
class AreaCalculator {
    public double calculateArea(List<Object> shapes, String type) {
        double area = 0.0;
        for (Object shape : shapes) {
            if (type.equals("rectangle") && shape instanceof Rectangle) {
                Rectangle rect = (Rectangle) shape;
                area += rect.width * rect.height;
            } else if (type.equals("circle") && shape instanceof Circle) {
                Circle circ = (Circle) shape;
                area += Math.PI * circ.radius * circ.radius;
            }
            // New shape types require modifying this method
        }
        return area;
    }
}

// Main class to demonstrate the functionality
public class OpenClosedPrinciple {
    public static void main(String[] args) {
        Rectangle rect = new Rectangle(5, 10);
        Circle circ = new Circle(7);

        List<Object> shapes = List.of(rect, circ);

        AreaCalculator calculator = new AreaCalculator();
        double totalArea = calculator.calculateArea(shapes, "rectangle") +
                           calculator.calculateArea(shapes, "circle");

        System.out.println("Total Area: " + totalArea);
    }
}

File: src/c04_professional_classes/matlab/OpenClosedPrinciple.m
function OpenClosedPrinciple()
    % In this example:
    % - AreaCalculator has to know about every shape type, checking the type string
    %   to calculate the area.
    % - This design violates the Open/Closed Principle because adding new shapes
    %   (e.g., Triangle) would require modifying calculateArea, leading to less
    %   maintainable code.

    rect = Rectangle(5, 10);
    circ = Circle(7);

    shapes = {rect, circ};

    calculator = AreaCalculator();
    totalArea = calculator.calculateArea(shapes, 'rectangle') + ...
                calculator.calculateArea(shapes, 'circle');

    fprintf('Total Area: %f\n', totalArea);
end

classdef Rectangle
    properties
        width
        height
    end
    methods
        function obj = Rectangle(width, height)
            obj.width = width;
            obj.height = height;
        end
    end
end

classdef Circle
    properties
        radius
    end
    methods
        function obj = Circle(radius)
            obj.radius = radius;
        end
    end
end

classdef AreaCalculator
    methods
        function area = calculateArea(obj, shapes, type)
            area = 0.0;
            for i = 1:length(shapes)
                shape = shapes{i};
                if strcmp(type, 'rectangle') && isa(shape, 'Rectangle')
                    area = area + shape.width * shape.height;
                elseif strcmp(type, 'circle') && isa(shape, 'Circle')
                    area = area + pi * shape.radius^2;
                end
                % New shape types require modifying this method
            end
        end
    end
end
